### PR TITLE
integration tests

### PR DESCRIPTION
**Pull request by fanatid**

### CODE REVIEW

**Comment by mappum** [comment]
Cool, good work so far :+1:
What's left to do for this?

BTW, can we download the binary from bitcoin.org instead of including it in the package? It's kind of scary to make people run an arbitrary binary. (Or at the very least, verify the hash of it matches the one on bitcoin.org).

**Comment by fanatid** [comment]
> What's left to do for this?

I want write tests for other messages.

> BTW, can we download the binary from bitcoin.org instead of including it in the package?

This binary from bitcoin.org, but people also can run their own bitcoind with environment variable [BITCOIND](https://github.com/fanatid/bitcoin-protocol/blob/470402bcf9fb5d8d7d84856119fcc0122f8561c4/test/integration/regtest.js#L28)

**Comment by fanatid** [comment]
@mappum ready for review and merge:
tested commands: version, verack, inv, getdata, getblocks, getheaders, block, headers, ping, pong
untested: addr, notfound, tx, getaddr, mempool, checkorder, submitorder, reply, reject, filterload, filteradd, filterclear, merkleblock, alter
maybe add them later

File: .travis.yml [hunk 1]
```diff
@@ -1,10 +1,19 @@
+sudo: false
 language: node_js
-
 node_js:
-- '5'
-- '4'
-
-notifications:
-  email: false
```
Comment (mappum) [inline .travis.yml]: Why turn email notifications back on? They get pretty annoying.

Comment (fanatid) [inline .travis.yml]: Sorry! I missed it. Returned back.

**Comment by mappum** [comment]
I tried running this, but I get some errors and then the tests hang:

```tap
TAP version 13
    # Subtest: Integration with bitcoin core in regtest mode
        # Subtest: spawn process
        not ok 1 - should be equal
          ---
          found:
            name: Error
            stack: >-
              Error: Method not found

                  at
              /home/mappum/bitcoin-protocol/node_modules/bitcoin/lib/jsonrpc.js:130:19

                  at Array.forEach (native)

                  at IncomingMessage.<anonymous>
              (/home/mappum/bitcoin-protocol/node_modules/bitcoin/lib/jsonrpc.js:127:15)

                  at emitNone (events.js:85:20)

                  at IncomingMessage.emit (events.js:179:7)

                  at endReadableNT (_stream_readable.js:906:12)

                  at nextTickCallbackWith2Args (node.js:475:9)

                  at process._tickCallback (node.js:389:17)
            message: Method not found
            code: -32601
          wanted: null
          compare: ===
          at:
            line: 98
            column: 11
            file: test/integration/regtest.js
          stack: |
            test/integration/regtest.js:98:11
            node_modules/bitcoin/lib/index.js:39:5
            node_modules/bitcoin/lib/jsonrpc.js:134:13
            IncomingMessage.<anonymous> (node_modules/bitcoin/lib/jsonrpc.js:127:15)
            endReadableNT (_stream_readable.js:906:12)
          source: |
            t.equal(err, null)
          ...
        1..1
        # failed 1 of 1 tests
    not ok 1 - spawn process # time=2575.17ms
      ---
      at:
        line: 63
        column: 5
        file: test/integration/regtest.js
      results:
        plan:
          start: 1
          end: 1
        count: 1
        pass: 0
        ok: false
        fail: 1
        time: 2575.17
      source: |
        t.test('spawn process', function (t) {
      ...

        # Subtest: handshake
        ok 1 - got data
        ok 2 - should be equal
        ok 3 - correct command
        ok 4 - correct version
        ok 5 - correct services
        ok 6 - correct timestamp
        ok 7 - got data
        ok 8 - should be equal
        ok 9 - correct command
        ok 10 - correct payload
        1..10
    ok 2 - handshake # time=27.648ms

        # Subtest: send ping -> wait pong
        ok 1 - got data
        ok 2 - should be equal
        ok 3 - correct command
        ok 4 - correct version
        ok 5 - correct services
        ok 6 - correct timestamp
        ok 7 - got data
        ok 8 - should be equal
        ok 9 - correct command
        ok 10 - correct payload
        ok 11 - expect truthy value
        ok 12 - should be equal
        1..12
    ok 3 - send ping -> wait pong # time=58.619ms

        # Subtest: wait inv -> send getdata -> wait block
        ok 1 - got data
        ok 2 - should be equal
        ok 3 - correct command
        ok 4 - correct version
        ok 5 - correct services
        ok 6 - correct timestamp
        ok 7 - got data
        ok 8 - should be equal
        ok 9 - correct command
        ok 10 - correct payload
        not ok 11 - should be equal
          ---
          found:
            name: Error
            stack: >-
              Error: Method not found

                  at
              /home/mappum/bitcoin-protocol/node_modules/bitcoin/lib/jsonrpc.js:130:19

                  at Array.forEach (native)

                  at IncomingMessage.<anonymous>
              (/home/mappum/bitcoin-protocol/node_modules/bitcoin/lib/jsonrpc.js:127:15)

                  at emitNone (events.js:85:20)

                  at IncomingMessage.emit (events.js:179:7)

                  at endReadableNT (_stream_readable.js:906:12)

                  at nextTickCallbackWith2Args (node.js:475:9)

                  at process._tickCallback (node.js:389:17)
            message: Method not found
            code: -32601
          wanted: null
          compare: ===
          at:
            line: 242
            column: 9
            file: test/integration/regtest.js
          stack: |
            test/integration/regtest.js:242:9
            node_modules/bitcoin/lib/index.js:39:5
            node_modules/bitcoin/lib/jsonrpc.js:134:13
            IncomingMessage.<anonymous> (node_modules/bitcoin/lib/jsonrpc.js:127:15)
            endReadableNT (_stream_readable.js:906:12)
          source: |
            t.equal(err, null)
          ...
        not ok 12 - expect truthy value
          ---
          at:
            line: 243
            column: 9
            file: test/integration/regtest.js
          stack: |
            test/integration/regtest.js:243:9
            node_modules/bitcoin/lib/index.js:39:5
            node_modules/bitcoin/lib/jsonrpc.js:134:13
            IncomingMessage.<anonymous> (node_modules/bitcoin/lib/jsonrpc.js:127:15)
            endReadableNT (_stream_readable.js:906:12)
          source: |
            t.ok(Array.isArray(data))
          ...
        not ok 13 - TypeError: Cannot read property 'length' of undefined
          ---
          at:
            line: 244
            column: 19
            file: test/integration/regtest.js
          stack: |
            test/integration/regtest.js:244:19
            node_modules/bitcoin/lib/index.js:39:5
            node_modules/bitcoin/lib/jsonrpc.js:134:13
            IncomingMessage.<anonymous> (node_modules/bitcoin/lib/jsonrpc.js:127:15)
            endReadableNT (_stream_readable.js:906:12)
          test: 'wait inv -> send getdata -> wait block'
          message: "TypeError: Cannot read property 'length' of undefined"
          source: |
            t.equal(data.length, 1)
          ...
        1..13
        # failed 3 of 13 tests
    not ok 4 - wait inv -> send getdata -> wait block # time=30.831ms
      ---
      at:
        line: 209
        column: 5
        file: test/integration/regtest.js
      results:
        plan:
          start: 1
          end: 13
        count: 13
        pass: 10
        ok: false
        fail: 3
        time: 30.831
      source: |
        t.test('wait inv -> send getdata -> wait block', connect(function (t) {
      ...

        # Subtest: send inv -> wait getdata -> send notfound
        ok 1 - got data
        ok 2 - should be equal
        ok 3 - correct command
        ok 4 - correct version
        ok 5 - correct services
        ok 6 - correct timestamp
        ok 7 - got data
        ok 8 - should be equal
        ok 9 - correct command
        ok 10 - correct payload
        1..10
    ok 5 - send inv -> wait getdata -> send notfound # time=7.637ms

        # Subtest: send getheaders -> wait headers
        ok 1 - got data
        ok 2 - should be equal
        ok 3 - correct command
        ok 4 - correct version
        ok 5 - correct services
        ok 6 - correct timestamp
        ok 7 - got data
        ok 8 - should be equal
        ok 9 - correct command
        ok 10 - correct payload
```

**Comment by fanatid** [comment]
@mappum can you try add `console.log(requestJSON)` in `node_modules/bitcoin/lib/jsonrpc.js`? What version of bitcoind are you using?

**Comment by mappum** [comment]
Ah, I was using Bitcoin Core 0.12, which removes support for the 'generate' method which was being requested. The tests work for me on 0.11.2. Thanks for adding this! :smile:

**Comment by fanatid** [comment]
Do you know what methods generate blocks in 0.12?

**Comment by mappum** [comment]
Hm, it looks like the `generate` method is still there: https://github.com/bitcoin/bitcoin/blob/0.12/src/rpcserver.cpp#L303

I'm not sure why I was getting 'Method not found' for calls to 'generate'.

**Pull request merged**